pkg/apis/pipeline/v1/pipeline_validation.go

```diff
@@ -184,13 +184,25 @@ func validatePipelineContextVariables(tasks []PipelineTask) *apis.FieldError {
 	var paramValues []string
 	for _, task := range tasks {
 		var matrixParams []Param
+		var includeParams []Param
 		if task.IsMatrixed() {
 			matrixParams = task.Matrix.Params
+			if task.Matrix.MatrixHasInclude() {
+				for _, include := range task.Matrix.Include {
+					includeParams = include.Params
+				}
+			}
 		}
 		for _, param := range append(task.Params, matrixParams...) {
 			paramValues = append(paramValues, param.Value.StringVal)
 			paramValues = append(paramValues, param.Value.ArrayVal...)
 		}
+
+		if task.Matrix.MatrixHasInclude() {
+			for _, param := range append(task.Params, includeParams...) {
+				paramValues = append(paramValues, param.Value.StringVal)
+			}
+		}
```

Comment:
We could leverage a common function retrieving values from params:

https://github.com/tektoncd/pipeline/pull/6180/files#diff-16493f3be05dfe9ce9d789d3b4b277f77d13deb8673084b3efe9034d20eb2987R121-R131

Also at this point in the execution, it must be safe to rely on this member function `Params.extractParamValuesFromParams`  which is collecting values from an objects as well since we have a validation upfront for limiting such usage.